[Bug]: one-sided AMM can rebase itself into its empty half

### Problem encountered

A one-sided AMM is able to get a position that puts it into its empty half during rebase on entry.

For example an AMM with no lower-bound (and so can only be short) but whose base is above the mid-price of the maket will submit a buy order to rebase its position so that it does not cross. The AMM then becomes long but has no curve definition on that side to evalulate. Mayhem ensues.

### Observed behaviour

```
panic: decimal division by 0

goroutine 114918 [running]:
github.com/shopspring/decimal.Decimal.QuoRem({0x14005c8cc20, 0xfffffff0}, {0x140003ad680, 0x1}, 0x10)
	/Users/caed/go/pkg/mod/github.com/vegaprotocol/decimal@v1.3.1-uint256/decimal.go:590 +0x244
github.com/shopspring/decimal.Decimal.DivRound({0x14005c8cc20?, 0x0?}, {0x140003ad680?, 0xfffffff0?}, 0x10)
	/Users/caed/go/pkg/mod/github.com/vegaprotocol/decimal@v1.3.1-uint256/decimal.go:632 +0x3c
github.com/shopspring/decimal.Decimal.Div(...)
	/Users/caed/go/pkg/mod/github.com/vegaprotocol/decimal@v1.3.1-uint256/decimal.go:577
code.vegaprotocol.io/vega/core/execution/amm.(*Pool).fairPrice(0x14002b3ce10)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/amm/pool.go:897 +0x104
code.vegaprotocol.io/vega/core/execution/amm.(*Pool).BestPrice(0x14002b3ce10, 0x0)
```

### Expected behaviour

No panic.

### Steps to reproduce

```bash
1. Submit `xyz` transaction
2. With this payload
3. Query `abc` API
4. Observe the error
```


### Software version

develop

### Failing test

_No response_

### Jenkins run

_No response_

### Configuration used

_No response_

### Relevant log output

_No response_

--- a/core/execution/amm/engine.go
+++ b/core/execution/amm/engine.go
@@ -720,13 +720,15 @@ func (e *Engine) Amend(
 		}
 	}
 
-	// we need to remove the existing pool from the engine so that when calculating rebasing orders we do not
-	// trade with ourselves.
-	e.remove(ctx, amend.Party)
 	updated, err := pool.Update(amend, riskFactors, scalingFactors, slippage)
 	if err != nil {
 		return nil, nil, err
 	}
+
+	// we need to remove the existing pool from the engine so that when calculating rebasing orders we do not
+	// trade with ourselves.
+	e.remove(ctx, amend.Party)
+
 	e.log.Debug("AMM amended",
 		logging.String("owner", amend.Party),
 		logging.String("marketID", e.marketID),

--- a/core/execution/amm/engine_test.go
+++ b/core/execution/amm/engine_test.go
@@ -60,7 +60,8 @@ func TestAMMTrading(t *testing.T) {
 
 func TestAmendAMM(t *testing.T) {
 	t.Run("test amend AMM which doesn't exist", testAmendAMMWhichDoesntExist)
-	t.Run("test amend AMM with sparse amend", TestAmendAMMSparse)
+	t.Run("test amend AMM with sparse amend", testAmendAMMSparse)
+	t.Run("test amend AMM insufficient commitment", testAmendInsufficientCommitment)
 }
 
 func TestClosingAMM(t *testing.T) {
@@ -103,7 +104,7 @@ func testAmendAMMWhichDoesntExist(t *testing.T) {
 	require.ErrorIs(t, err, ErrNoPoolMatchingParty)
 }
 
-func TestAmendAMMSparse(t *testing.T) {
+func testAmendAMMSparse(t *testing.T) {
 	ctx := context.Background()
 	tst := getTestEngine(t)
 
@@ -127,6 +128,33 @@ func TestAmendAMMSparse(t *testing.T) {
 	require.NoError(t, err)
 
 	tst.engine.Confirm(ctx, updated)
+}
+
+func testAmendInsufficientCommitment(t *testing.T) {
+	ctx := context.Background()
+	tst := getTestEngine(t)
+
+	party, subAccount := getParty(t, tst)
+	submit := getPoolSubmission(t, party, tst.marketID)
+	expectSubaccountCreation(t, tst, party, subAccount)
+	whenAMMIsSubmitted(t, tst, submit)
+
+	poolID := tst.engine.poolsCpy[0].ID
+
+	amend := getPoolAmendment(t, party, tst.marketID)
+	// no amend to the commitment amount
+	amend.CommitmentAmount = nil
+
+	// amend to super wide bounds so that the commitment is too thin to support the AMM
+	amend.Parameters.Base.AddSum(num.UintOne())
+	amend.Parameters.UpperBound.AddSum(num.NewUint(1000000))
+	amend.Parameters.LowerBound.AddSum(num.UintOne())
+
+	_, _, err := tst.engine.Amend(ctx, amend, riskFactors, scalingFactors, slippage)
+	require.ErrorContains(t, err, "insufficient commitment")
+
+	// check that the original pool still exists
+	assert.Equal(t, poolID, tst.engine.poolsCpy[0].ID)
 }
 
 func testBasicSubmitOrder(t *testing.T) {

--- a/core/execution/amm/pool.go
+++ b/core/execution/amm/pool.go
@@ -891,6 +891,10 @@ func (p *Pool) fairPrice() *num.Uint {
 		pv = cu.pv.Add(pv)
 	}
 
+	if cu.empty {
+		panic("should not be calculating fair-price on empty-curve side")
+	}
+
 	l := cu.l
 
 	// pv * sqrt(pu) * (1/L) + 1

